Libs/MRML/Core/vtkMRMLNode.cxx

```diff
@@ -368,52 +368,60 @@ void vtkMRMLNode::ReadXMLAttributes(const char** atts)
   vtkMRMLReadXMLEndMacro();
 
   std::map<std::string, std::string> references;
-
+  const char** xmlReadAtts = atts;
   const char* attName;
   const char* attValue;
-  while (*atts != nullptr)
-    {
-    attName = *(atts++);
-    attValue = *(atts++);
-
-     if (!strcmp(attName, "attributes"))
-       {
-       std::stringstream attributes(attValue);
-       std::string attribute;
-       while (std::getline(attributes, attribute, ';'))
-         {
-         int colonIndex = attribute.find(':');
-         std::string name = attribute.substr(0, colonIndex);
-         std::string value = attribute.substr(colonIndex + 1);
-         // decode percent sign and semicolon (semicolon is a special character because it separates attributes)
-         vtksys::SystemTools::ReplaceString(value, "%3B", ";");
-         vtksys::SystemTools::ReplaceString(value, "%25", "%");
-         this->SetAttribute(name.c_str(), value.c_str());
-         }
-       }
-     else if (!strcmp(attName, "references"))
-       {
-       this->ParseReferencesAttribute(attValue, references);
-       }
-     else if ( const char* referenceRole =
-                 this->GetReferenceRoleFromMRMLAttributeName(attName) )
-       {
-       std::stringstream ss(attValue);
-       while (!ss.eof())
-         {
-         std::string id;
-         ss >> id;
-         if (!id.empty())
-           {
-           if (references.find(id) == references.end() ||
-             references.find(id)->second != referenceRole)
```

Comment:
The same node may be referenced with different roles (e.g., in SlicerRT DVH plot, the same nodes are referenced using DoseVolumeHistogram.dvhPlotSeries and plotSeries role). In such cases the logic is still not correct because in "references" variable we only store the last encountered role name.

We should just keep a list of roles (can be a std::set<std::string>) and if we find that a reference role has been encountered before then we don't add any nodes with it again.